Homework/lesson 17/task1.py

```diff
@@ -0,0 +1,202 @@
+# # task1
+# class Animal:
+#     def __init__(self, name):
+#         self.name = name
+#
+#     def talk(self):
+#         raise NotImplementedError("Give me animal")
+#
+#
+# class Cat(Animal):
+#     def talk(self):
+#         print("Meow")
+#
+#
+# class Dog(Animal):
+#     def talk(self):
+#         print("Gaw, Gaw")
+#
+#
+# def make_animal_talk(animals):
+#     animals.talk(Animal)
+#
+#
+# dog = Dog("Ali")
+# cat = Cat("Tyson")
+#
+# make_animal_talk(Dog)
+# make_animal_talk(Cat)
+#
+# animals = [Cat("Tyson"), Dog("Ali")]
+# for animal in animals: animal.talk()
+
+
+# task2
+class Library:
+    def __init__(self, name: str):
+        self.name = name
+        self.books = []
+        self.authors = []
+
+    def new_book(self, name: str, year: int, author: 'Author') -> 'Book':
+        book = Book(name, year, author)
+        self.books.append(book)
+        self.authors.append(author)
+        return book
+
+    def group_by_author(self, author):
+        author_books = [book for book in self.books if book.author == author]
+        return author_books
+
+    def group_by_year(self, year: int) -> list:
+        year_books = [book for book in self.books if book.year == year]
+        return year_books
+
+    def __str__(self):
+        return f"Library({self.name}, books={self.books}, authors={self.authors})"
+
+    def __repr__(self):
+        return self.name
+
+
+class Book:
+    total_books = 0
+
+    def __init__(self, name, year, author):
+        self.name = name
+        self.year = year
+        self.author = author
+        self.author.books.append(self)
+        Book.total_books += 1
+
+    def __str__(self):
+        return f"Library({self.name}, {self.year}, authors={self.author})"
+
+    def __repr__(self):
+        return self.name
+
+
+class Author:
+    def __init__(self, name, country, birthday):
+        self.name = name
+        self.country = country
+        self.birthday = birthday
+        self.books = []
+
+    def add_book(self, book):
+        self.books.append(book)
+
+    def __str__(self):
+        return f"Library({self.name}, {self.country}, {self.birthday}, {self.books})"
+
+    def __repr__(self):
+        return self.name
+
+
+if __name__ == "__main__":
+    # Create some authors
+    Dumas = Author("Alexandre Dumas ", "France", "24 July 1802")
```

Comment:
It's better to name variables in lowercase. Even for names...
```suggestion
    dumas = Author("Alexandre Dumas ", "France", "24 July 1802")
```

Reply:
Gotcha